packages/astro/src/vite-plugin-markdown/index.ts

```diff
@@ -32,247 +23,95 @@ function safeMatter(source: string, id: string) {
-${isAstroFlavoredMd ? setup : ''}`.trim();
+				const code = escapeViteEnvReferences(`
+				import { Fragment, jsx as h } from 'astro/jsx-runtime';
+				${layout ? `import Layout from ${JSON.stringify(layout)};` : ''}
 
-				// Wrap with set:html fragment to skip
-				// JSX expressions and components in "plain" md mode
-				if (!isAstroFlavoredMd) {
-					astroResult = `<Fragment set:html={${JSON.stringify(astroResult)}} />`;
+				export const frontmatter = ${JSON.stringify(frontmatter)};
+				export const file = ${JSON.stringify(fileId)};
+				export const url = ${JSON.stringify(fileUrl)};
+				export function rawContent() {
+					return ${JSON.stringify(raw.content)};
 				}
-
-				// If the user imported "Layout", wrap the content in a Layout
-				if (/\bLayout\b/.test(imports)) {
-					astroResult = `${prelude}\n<Layout content={$$content}>\n\n${astroResult}\n\n</Layout>`;
-				} else {
-					// Note: without a Layout, we need to inject `head` manually so `maybeRenderHead` runs
-					astroResult = `${prelude}\n<head></head>${astroResult}`;
+				export function compiledContent() {
+					return ${JSON.stringify(html)};
 				}
-
-				// Transform from `.astro` to valid `.ts`
-				const compileProps: CompileProps = {
-					config,
-					filename,
-					moduleId: id,
-					source: astroResult,
-					ssr: Boolean(opts?.ssr),
-					viteTransform,
-					pluginContext: this,
-				};
-
-				let transformResult = await cachedCompilation(compileProps);
-				let { code: tsResult } = transformResult;
-
-				tsResult = `\nexport const metadata = ${JSON.stringify(metadata)};
-export const frontmatter = ${JSON.stringify(content)};
-export function rawContent() {
-	return ${JSON.stringify(markdownContent)};
-}
-export function compiledContent() {
-		return ${JSON.stringify(renderResult.metadata.html)};
-}
-${tsResult}`;
-
-				// Compile from `.ts` to `.js`
-				const { code } = await esbuild.transform(tsResult, {
-					loader: 'ts',
-					sourcemap: false,
-					sourcefile: id,
-				});
-
-				const astroMetadata: AstroPluginMetadata['astro'] = {
-					clientOnlyComponents: transformResult.clientOnlyComponents,
-					hydratedComponents: transformResult.hydratedComponents,
-					scripts: transformResult.scripts,
+				export function getHeadings() {
+					return ${JSON.stringify(headings)};
+				}
+				export function getHeaders() {
+					console.warn('getHeaders() have been deprecated. Use getHeadings() function instead.');
+					return getHeadings();
 				};
-
+				export async function Content() {
+					const { layout, ...content } = frontmatter;
+					content.astro = {};
+					Object.defineProperty(content.astro, 'headings', {
+						get() {
+							throw new Error('The "astro" property is no longer supported! To access "headings" from your layout, try using "Astro.props.headings."')
+						}
+					});
+					const contentFragment = h(Fragment, { 'set:html': ${JSON.stringify(html)} });
```

Comment:
I just realized that you're re-calling `JSON.stringify` on the same html/raw values multiple times per file which probably does have a perf impact! 

You could either move these to variables in the plugin context, so that you're running `JSON.stringify` once but inlining the string multiple times in the file itself. OR, you could move these into top-level variables in the file itself, so that `JSON.stringify` is only called once AND the string only exists once in the file.

I'd love for you to explore the later one. I bet my pride that this has a noticeable impact on the DDT benchmark!

Reply:
Made a top-level variable! 👍